Assets/SEE/GraphProviders/MultiGraphPipelineProvider.cs

```diff
@@ -0,0 +1,128 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Cysharp.Threading.Tasks;
+using SEE.DataModel.DG;
+using SEE.Game.City;
+using SEE.Utils.Config;
+using Sirenix.OdinInspector;
+
+namespace SEE.GraphProviders
+{
+    /// <summary>
+    /// A graph provider pipeline of multiple <see cref="MultiGraphProvider"/>
+    /// </summary>
+    public class MultiGraphPipelineProvider : MultiGraphProvider
+    {
+        /// <summary>
+        /// The list of nested providers in this pipeline. These will be executed
+        /// from first to last.
+        /// </summary>
+        [HideReferenceObjectPicker,
+         ListDrawerSettings(DefaultExpandedState = true, ListElementLabelName = nameof(GetKind))]
+        public List<MultiGraphProvider> Pipeline = new();
+
+        /// <summary>
+        /// Provides a graph based as a result of the serial execution of all
+        /// graph providers in <see cref="Pipeline"/> from first to last.
+        ///
+        /// If the <see cref="Pipeline"/> is empty, <paramref name="graphs"/>
+        /// will be returned.
+        /// </summary>
+        /// <param name="graphs">this list of graphs will be input to the first
+        /// graph provider in <see cref="Pipeline"/></param>
+        /// <param name="city">this value will be passed to each graph provider
+        /// in <see cref="Pipeline"/></param>
+        /// <param name="changePercentage">this callback will be called with
+        /// the percentage (0–1) of completion of the pipeline</param>
+        /// <param name="token">can be used to cancel the operation</param>
+        /// <returns></returns>
+        /// <remarks>Exceptions may be thrown by each nested graph provider.</remarks>
+        public override async UniTask<List<Graph>> ProvideAsync(List<Graph> graphs, AbstractSEECity city,
+            Action<float> changePercentage = null,
+            CancellationToken token = default)
+        {
+            UniTask<List<Graph>> initial = UniTask.FromResult(graphs);
+            // Counts the number of providers that have been executed so far.
+            // Initially -1 because the first provider will increment it to 0.
+            int count = -1;
+
+            return await Pipeline.Aggregate(initial, (current, provider) =>
+                current.ContinueWith(g => provider.ProvideAsync(g, city, AggregatePercentage(), token)));
+
+            Action<float> AggregatePercentage()
+            {
+                count++;
+                // Each stage of the pipeline gets an equal share of the total percentage.
+                return percentage => changePercentage?.Invoke((count + percentage) / Pipeline.Count);
+            }
```

Comment:
This seems to clone from `SingleGraphPipelineProvider`. Is there any possibility to extract this common functionality (e.g., percentage aggregation) or would that overcomplicate the code?

Reply:
I eleminated the redundant code including the saving and restoring of the pipeline, which were also copied and pasted.